src/Training.AirBnb.Clone.Backend/Backend Project/Domain/Services/CountryService.cs

```diff
@@ -0,0 +1,119 @@
+            if (!IsValdCountryName(country))
+                throw new CountryFormatException("The Country is in the wrong format");
+            if (!(IsValidCountryDailingCode(country)) && !(IsValidRegionPhoneNumberLength(country)))
+                throw new CountryFormatException("This Country is in the wrong format");
+
+            await _appDataContext.Countries.AddAsync(country);
+            if (saveChanges)
+                await _appDataContext.SaveChangesAsync();
+            return country;
+        }
+
+        public async ValueTask<Country> DeleteAsync(Guid id, bool saveChanges = true)
+        {
+            var deletedCountry = await GetByIdAsync(id);
+            if (deletedCountry is null)
+                throw new CountryNotFoundException("Country not found");
+            deletedCountry.DeletedDate = DateTimeOffset.UtcNow;
+            deletedCountry.IsDeleted = true;
+            if (saveChanges)
+                await _appDataContext.Countries.SaveChangesAsync();
+            return deletedCountry;
+        }
+
+        public async ValueTask<Country> DeleteAsync(Country country, bool saveChanges = true)
+        {
+            var deletedCountry = await GetByIdAsync(country.Id);
+            if (deletedCountry is null)
+                throw new CountryNotFoundException("this Country not found");
+            deletedCountry.DeletedDate = DateTimeOffset.UtcNow;
+            deletedCountry.IsDeleted = true;
+            if (saveChanges)
+                await _appDataContext.Countries.SaveChangesAsync();
+            return deletedCountry;
+        }
+
+        public IQueryable<Country> Get(Expression<Func<Country, bool>> predicate)
+        {
+            return GetUndeletedCountries().Where(predicate.Compile()).AsQueryable();
+        }
+
+        public ValueTask<ICollection<Country>> GetAsync(IEnumerable<Guid> ids)
+        {
+            var countries = GetUndeletedCountries()
+            .Where(country => ids.Contains(country.Id));
+            return new ValueTask<ICollection<Country>>(countries.ToList());
+        }
+
+        public ValueTask<Country> GetByIdAsync(Guid id)
+        {
+            var foundCountry = GetUndeletedCountries().FirstOrDefault(country => country.Id.Equals(id));
+            if (foundCountry is null)
+                throw new CountryNotFoundException("Country not found");
+            return new ValueTask<Country>(foundCountry);
+        }
+
+        public async ValueTask<Country> UpdateAsync(Country country, bool saveChanges = true)
+        {
+            var foundCountry = await GetByIdAsync(country.Id);
+            if (foundCountry is null)
+                throw new CountryNotFoundException("Country not found");
+            if (!IsValdCountryName(country))
+                throw new CountryFormatException("The Country is in the wrong format");
+            if (!(IsValidCountryDailingCode(country)) && !(IsValidRegionPhoneNumberLength(country)))
```

Comment:
ikkita methoda ham false qaytsagina exception bo'lyapti, lekin har ikkalasidan bittasida false qaytsa exception bo'lishi kerak, yani or ishlatish kerak and o'rniga.